dss/events/__init__.py

```diff
@@ -0,0 +1,129 @@
+    """
+    Build the bundle metadata document, record it into flashflood, and return it
+    """
+    # TODO: Add support for unversioned tombstones
+    fqid = key.split("/", 1)[1]
+    ff = FlashFlood(resources.s3,  # type: ignore
+                    Config.get_flashflood_bucket(),
+                    flashflood_prefix or replica.flashflood_prefix)
+    metadata_document = _build_bundle_metadata_document(replica, key)
+    if use_version_for_timestamp:
+        _, version = fqid.split(".", 1)
+        event_date = datetime_from_timestamp(version)
+    else:
+        event_date = datetime.utcnow()
+    ff.put(json.dumps(metadata_document).encode("utf-8"), event_id=fqid, date=event_date)
+    return metadata_document
+
+def _build_bundle_metadata_document(replica: Replica, key: str) -> dict:
+    """
+    This returns a JSON document with bundle manifest and metadata files suitable for JMESPath filters.
+    """
+    handle = Config.get_blobstore_handle(replica)
+    manifest = json.loads(handle.get(replica.bucket, key).decode("utf-8"))
+    if key.endswith(TOMBSTONE_SUFFIX):
+        manifest['event_type'] = "TOMBSTONE"
+        return manifest
+    else:
+        lock = threading.Lock()
+        files: dict = defaultdict(list)
+
+        def _read_file(file_metadata):
+            blob_key = "blobs/{}.{}.{}.{}".format(
+                file_metadata['sha256'],
+                file_metadata['sha1'],
+                file_metadata['s3-etag'],
+                file_metadata['crc32c'],
+            )
+            contents = handle.get(replica.bucket, blob_key).decode("utf-8")
+            try:
+                file_info = json.loads(contents)
+            except json.decoder.JSONDecodeError:
+                logging.info(f"{file_metadata['name']} not json decodable")
+            else:
+                # Modify name to avoid confusion with JMESPath syntax
+                name = _dot_to_underscore_and_strip_numeric_suffix(file_metadata['name'])
+                with lock:
+                    files[name].append(file_info)
+
+        # TODO: Consider scaling parallelization with Lambda size
+        with ThreadPoolExecutor(max_workers=20) as e:
+            e.map(_read_file, [file_metadata for file_metadata in manifest['files']
+                               if file_metadata['content-type'].startswith("application/json")])
+
+        return {
+            'event_type': "CREATE",
+            'manifest': manifest,
+            'files': dict(files),
+        }
+
+def _dot_to_underscore_and_strip_numeric_suffix(name: str) -> str:
+    """
+    e.g. "library_preparation_protocol_0.json" -> "library_preparation_protocol_json"
+    """
+    name = name.replace('.', '_')
+    if name.endswith('_json'):
+        name = name[:-5]
+        parts = name.rpartition("_")
+        if name != parts[2]:
+            name = parts[0]
+        name += "_json"
+    return name
+
+def datetime_from_timestamp(ts):
```

Comment:
nit: keep this loaded in `dss.util.version` 🔫

Reply:
done